### ISSUE TITLE
`map_union` returns different results under fuzzer

### ISSUE THREAD

**Issue by Yuhta** [facebookincubator/velox#3402]
### Description

https://app.circleci.com/pipelines/github/facebookincubator/velox/17149/workflows/05f88c81-861b-4aab-8631-57e918dc2462/jobs/107154

### Error Reproduction

Run aggregation fuzzer with seed 1531098586 and version 1eaa9a475f8e6af1bafab6b52674f08e0e44e227

```
#!/bin/bash -eo pipefail
echo -e "The logs will be saved to \"/tmp/aggregate_fuzzer.log\" and uploaded" \
        "as a job artifact (check the \"Artifacts\" tab above)."
_build/debug/velox/exec/tests/velox_aggregation_fuzzer_test \
    --seed ${RANDOM} \
    --duration_sec 3600 \
    --logtostderr=1 \
    --minloglevel=0 \
  2>&1 | tee /tmp/aggregate_fuzzer.log || ( \
    tail -n 1000 /tmp/aggregate_fuzzer.log;
    echo "FAIL: Aggregate Fuzzer run failed";
    exit 1; \
  )
echo -e "\n\nAggregate Fuzzer run finished successfully."
```

### Relevant logs

```bash
I1202 03:33:54.807240  7735 AggregationFuzzer.cpp:481] ==============================> Started iteration 1166 (seed: 1531098586)
I1202 03:33:54.813868  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[SINGLE [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- Values[1000 rows in 10 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.820756  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
I1202 03:33:54.844281  7735 AggregationFuzzer.cpp:170] Testing plan #0
I1202 03:33:54.844306  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[FINAL [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
      -- Values[1000 rows in 10 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.853734  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
I1202 03:33:54.860131  7735 AggregationFuzzer.cpp:173] Testing plan #0 with spilling
I1202 03:33:54.860154  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[FINAL [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
      -- Values[1000 rows in 10 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.869805  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
I1202 03:33:54.876148  7735 AggregationFuzzer.cpp:170] Testing plan #1
I1202 03:33:54.876170  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[FINAL [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- Aggregation[INTERMEDIATE [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[1000 rows in 10 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.888319  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
I1202 03:33:54.894273  7735 AggregationFuzzer.cpp:173] Testing plan #1 with spilling
I1202 03:33:54.894294  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[FINAL [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- Aggregation[INTERMEDIATE [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[1000 rows in 10 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.906445  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
I1202 03:33:54.912762  7735 AggregationFuzzer.cpp:170] Testing plan #2
I1202 03:33:54.912782  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[FINAL [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- LocalPartition[REPARTITION] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[300 rows in 3 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[300 rows in 3 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[200 rows in 2 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[200 rows in 2 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.918612  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
I1202 03:33:54.924816  7735 AggregationFuzzer.cpp:173] Testing plan #2 with spilling
I1202 03:33:54.924837  7735 AggregationFuzzer.cpp:561] Executing query plan: 
-- Project[expressions: (g0:DOUBLE, ROW["g0"]), (p1:ARRAY<TINYINT>, array_sort(map_keys(ROW["a0"])))] -> g0:DOUBLE, p1:ARRAY<TINYINT>
  -- Aggregation[FINAL [g0] a0 := map_union("a0")] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
    -- LocalPartition[REPARTITION] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[300 rows in 3 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[300 rows in 3 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[200 rows in 2 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
      -- Aggregation[PARTIAL [g0] a0 := map_union(ROW["c0"]) mask: m0] -> g0:DOUBLE, a0:MAP<TINYINT,TINYINT>
        -- Values[200 rows in 2 vectors] -> c0:MAP<TINYINT,TINYINT>, m0:BOOLEAN, g0:DOUBLE
I1202 03:33:54.933530  7735 AggregationFuzzer.cpp:576] [ROW ROW<g0:DOUBLE,p1:ARRAY<TINYINT>>: 483 elements, no nulls]
../../velox/exec/tests/utils/QueryAssertions.cpp:986: Failure
Value of: false
  Actual: false
Expected: true
Expected 483, got 483
1 extra rows, 1 missing rows
1 of extra rows:
	null | [0,1,2,4,5,5,6,7,8,9,10,12,13,16,17,18,19,20,21,22,23,23,24,25,26,27,27,28,28,29,30,31,32,33,33,34,35,35,36,37,40,41,42,43,44,45,46,48,49,50,52,53,54,55,56,59,64,65,66,67,68,69,71,73,75,75,76,76,77,78,79,80,81,82,83,84,85,87,88,89,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,107,108,109,110,110,112,113,114,115,118,120,121,122,123,125,126,127]

1 of missing rows:
	null | [0,1,2,4,5,6,7,8,9,10,12,13,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,40,41,42,43,44,45,46,48,49,50,52,53,54,55,56,59,64,65,66,67,68,69,71,73,75,76,77,78,79,80,81,82,83,84,85,87,88,89,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,112,113,114,115,118,120,121,122,123,125,126,127]

Unexpected results
E1202 03:33:54.940344  7735 Exceptions.h:68] Line: ../../velox/exec/tests/AggregationFuzzer.cpp:786, Function:testPlan, Expression: assertEqualResults({expected.result}, {actual.result}) Logically equivalent plans produced different results, Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: Logically equivalent plans produced different results
Retriable: False
Expression: assertEqualResults({expected.result}, {actual.result})
Function: testPlan
File: ../../velox/exec/tests/AggregationFuzzer.cpp
Line: 786
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxException5State4makeIZNS1_C4EPKcmS5_St17basic_string_viewIcSt11char_traitsIcEES9_S9_S9_bNS1_4TypeES9_EUlRT_E_EESt10shared_ptrIKS2_ESA_SB_
# 2  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 3  _ZN8facebook5velox17VeloxRuntimeErrorC2EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bS7_
# 4  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorEPKcEEvRKNS1_18VeloxCheckFailArgsET0_
# 5  _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer8testPlanERKSt10shared_ptrIKNS0_4core8PlanNodeEEbbRKNS3_13ResultOrErrorE
# 6  _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer9testPlansERKSt6vectorISt10shared_ptrIKNS0_4core8PlanNodeEESaISA_EEbRKNS3_13ResultOrErrorE
# 7  _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer6verifyERKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaISB_EESF_SF_RKS5_ISt10shared_ptrINS0_9RowVectorEESaISI_EEbSF_
# 8  _ZN8facebook5velox4exec4test12_GLOBAL__N_117AggregationFuzzer2goEv
# 9  _ZN8facebook5velox4exec4test15aggregateFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorISt10shared_ptrINS1_26AggregateFunctionSignatureEESaISD_EESt4hashIS9_ESt8equal_toIS9_ESaISt4pairIKS9_SF_EEEmRKS3_IS9_S9_SH_SJ_SaISK_ISL_S9_EEE
# 10 _ZN23AggregationFuzzerRunner3runERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEmRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EERKSt13unordered_mapIS5_S5_SA_SC_SaISt4pairIS6_S5_EEE
# 11 main
# 12 __libc_start_main
# 13 _start

*** Aborted at 1669952035 (Unix time, try 'date -d @1669952035') ***
*** Signal 6 (SIGABRT) (0x1e37) received by PID 7735 (pthread TID 0x7fa1a8cc4380) (linux TID 7735) (maybe from PID 7735, UID 0) (code: -6), stack trace: ***
(error retrieving stack trace)
FAIL: Aggregate Fuzzer run failed
```

**Comment by mbasmanova** [comment]
Looks like there are duplicate keys: `5,5`

**Comment by mbasmanova** [comment]
This issue might have been fixed in #3291 by resetting sortedKeys_ in MapVector::prepareForReuse.

```
void MapVector::prepareForReuse() {
  BaseVector::prepareForReuse();

  sortedKeys_ = false;
```

**Comment by mbasmanova** [comment]
Confirmed that this is fixed in #3291
